runtime/src/message_processor.rs

```diff
@@ -66,38 +66,39 @@ fn verify_instruction(
 ) -> Result<(), InstructionError> {
     // Verify the transaction
 
-    // Make sure that program_id is still the same or this was just assigned by the system program,
-    //  but even the system program can't touch a credit-only account.
-    if pre.owner != post.owner && (!is_debitable || !system_program::check_id(&program_id)) {
+    // Make sure that program_id is still the same or this was just re-assigned by
+    //  the owner program
+    if pre.owner != post.owner && (!is_debitable || *program_id != pre.owner) {
         return Err(InstructionError::ModifiedProgramId);
     }
     // An account not assigned to the program cannot have its balance decrease.
-    if *program_id != post.owner && pre.lamports > post.lamports {
+    if *program_id != pre.owner && pre.lamports > post.lamports {
         return Err(InstructionError::ExternalAccountLamportSpend);
     }
     // The balance of credit-only accounts may only increase.
     if !is_debitable && pre.lamports > post.lamports {
         return Err(InstructionError::CreditOnlyLamportSpend);
     }
     // Only system accounts can change the size of the data.
-    if !system_program::check_id(&program_id) && pre.data.len() != post.data.len() {
+    if pre.data.len() != post.data.len() && !system_program::check_id(program_id) {
         return Err(InstructionError::AccountDataSizeChanged);
     }
-    // For accounts unassigned to the program, the data may not change.
-    if *program_id != post.owner && !system_program::check_id(&program_id) && pre.data != post.data
-    {
-        return Err(InstructionError::ExternalAccountDataModified);
-    }
-    // Credit-only account data may not change.
-    if !is_debitable && pre.data != post.data {
-        return Err(InstructionError::CreditOnlyDataModified);
+    // Verify data...
+    if pre.data != post.data {
+        // Credit-only account data may not change.
+        if !is_debitable {
+            return Err(InstructionError::CreditOnlyDataModified);
+        }
+        // For accounts not assigned to the program, the data may not change.
+        if *program_id != pre.owner && !system_program::check_id(program_id) {
```

Comment:
I'm curious about this pseudo-restriction, too.  it looks to me like we're actually being more permissive by allowing the system program to change a program's data